verus/lcf/src/verusport.rs

```diff
@@ -0,0 +1,1119 @@
+        requires
+            self.deep_view().spec_complete_subst(s.deep_view()),
+        ensures
+            res.deep_view().spec_concrete(),
+            res.deep_view() == self.deep_view().spec_subst(s.deep_view()),
+    {
+        match self {
+            Term::Var(v) => {
+                let u_option = s.get(v.as_str());
+                Term::Const(u_option.unwrap().clone())
+            },
+            Term::Const(_) => self,
+        }
+    }
+}
+
+//#[verifier::external_body] needed because the verifier does not support the equality relation we are trying to leverage
+#[verifier::external_body]
+//function checks whether 2 Vec<Term>s are equal or not
+pub fn terms_eq(a: &Vec<Term>, b: &Vec<Term>) -> (res: bool)
+    ensures
+        res <==> a.deep_view() == b.deep_view(),
+{
+    a == b
+}
+
+pub enum Prop {
+    App(String, Vec<Term>),
+    Eq(Term, Term),
+}
+
+pub enum SpecProp {
+    App(Seq<char>, Seq<SpecTerm>),
+    Eq(SpecTerm, SpecTerm),
+}
+
+impl DeepView for Prop {
+    type V = SpecProp;
+
+    // deep_view allows for reasoning about spec-level data structures and types while in exec mode
+    open spec fn deep_view(&self) -> Self::V {
+        match self {
+            Prop::App(s, v) => SpecProp::App(s.view(), v.deep_view()),
+            Prop::Eq(t, e) => SpecProp::Eq(t.deep_view(), e.deep_view()),
+        }
+    }
+}
+
+impl Clone for Prop {
+    //#[verifier::external_body] needed to satisfy postcondition on line 253.
+    #[verifier::external_body]
+    //clone function for Prop types explicitly copies and instantiates the original value into a new item so that it fixes borrowing errors.
+    //It is needed when you want to borrow a value that will produce an error when you try to
+    fn clone(&self) -> (res: Self)
+        ensures
+            self == res,
+    {
+        match self {
+            Prop::App(s, v) => Prop::App(s.clone(), v.clone()),
+            Prop::Eq(t, e) => Prop::Eq(t.clone(), e.clone()),
+        }
+    }
+}
+
+impl PartialEq for Prop {
+    // function aids verus verifier in reasoning about equality; not fully sufficient - must call prop_eq for certain types
+    fn eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Prop::App(s, v), Prop::App(t, w)) => s == t && v == w,
+            (Prop::Eq(a, b), Prop::Eq(c, d)) => a == c && b == d,
+            (Prop::App(_, _), Prop::Eq(_, _)) | (Prop::Eq(_, _), Prop::App(_, _)) => false,
+        }
+    }
+}
+
+impl SpecProp {
+    //spec function checks if Prop variants contain key to map; calls spec_complete_subst from Term
+    pub open spec fn spec_complete_subst(self, s: SpecSubst) -> bool {
+        match self {
+            SpecProp::App(head, args) => forall|i: int|
+                #![auto]
+            //forall checks every index in list of SpecTerms in args to check if they are complete substitutions
```

Comment:
Nitpicking: By placing this comment here, after the `forall`, the formatter seems to be adding this extra newline and making it somewhat harder to read. If you want a comment about the quantifier, I'd put it on the line above the quantifier. But in this case I don't think it's needed---it should be obvious to Verus users what a `forall` quantifier does.